pkg/controller/volume/persistentvolume/pv_controller.go

```diff
@@ -1342,13 +1421,45 @@ func (ctrl *PersistentVolumeController) provisionClaim(claim *v1.PersistentVolum
 	}
 	klog.V(4).Infof("provisionClaim[%s]: started", claimToClaimKey(claim))
 	opName := fmt.Sprintf("provision-%s[%s]", claimToClaimKey(claim), string(claim.UID))
-	startTime := time.Now()
-	ctrl.scheduleOperation(opName, func() error {
-		pluginName, err := ctrl.provisionClaimOperation(claim)
-		timeTaken := time.Since(startTime).Seconds()
-		metrics.RecordVolumeOperationMetric(pluginName, "provision", timeTaken, err)
-		return err
-	})
+	plugin, storageClass, err := ctrl.findProvisionablePlugin(claim)
+	if err != nil {
+		ctrl.eventRecorder.Event(claim, v1.EventTypeWarning, events.ProvisioningFailed, err.Error())
+		klog.Errorf("error finding provisioning plugin for claim %s: %v", claimToClaimKey(claim), err)
+		// failed to find the requested provisioning plugin, directly return err for now.
+		// controller will retry the provisioning in every syncUnboundClaim() call
+		// retain the original behavior of returning nil from provisionClaim call
+		return nil
+	}
+	// for CIS provisioning, latency metric will be reported after final binding happened
+	if plugin == nil || plugin.IsMigratedToCSI() {
+		// TEST case
+		ctrl.scheduleOperation(opName, func() error {
+			// create an opTimestamps without assigning provisionerName to get the current timestamp
+			opTimestamps := newOperationTimestamp("")
+			provisionerName, err := ctrl.provisionClaimOperationCSI(claim, plugin, storageClass)
+			opTimestamps.provisionerName = provisionerName
+			// adding an entry into operationTimestamps for metrics reporting if there exists no
+			// such record for current claim regardless whether the operation returned error or not
+			// add an entry to mark the provisioning request start time
+			claimKey := claimToClaimKey(claim)
+			provisionStartTsKey := ctrl.createOperationTimestampCacheKey("provision", claimKey)
+			if _, exists := ctrl.operationTimestamps.Get(provisionStartTsKey); !exists {
+				ctrl.operationTimestamps.Add(provisionStartTsKey, opTimestamps)
+			}
```

Comment:
Can  a user cause unlimited memory growth, knowing PVC operations for which CSI driver does not exist will not be cleaned up? Do we need a periodic cleanup to ensure cache does not grow out of bounds?

Reply:
answered above